packages/core/src/sagas/connectSaga.js

```diff
@@ -122,74 +133,129 @@ function* connectSaga( directLine ) {
         || connectionStatus === FAILED_TO_CONNECT
       ) {
         // If we receive anything negative, we will assume the connection is errored out
-        throw new Error( 'Failed to connect' );
+        throw new Error('Failed to connect');
       }
     }
   } finally {
-    if ( yield cancelled() ) {
+    if (yield cancelled()) {
       activitySubscription.unsubscribe();
 
-      throw new Error( 'Cancelled' );
+      throw new Error('Cancelled');
     }
   }
 }
 
 export default function* () {
-  for ( ; ; ) {
+  for (;;) {
     const {
       payload: {
         directLine,
         userID: userIDFromAction,
         username
       }
-    } = yield take( CONNECT );
-    const updateConnectionStatusTask = yield fork( observeAndPutConnectionStatusUpdate, directLine );
+    } = yield take(CONNECT);
+
+    const updateConnectionStatusTask = yield fork(observeAndPutConnectionStatusUpdate, directLine);
 
     try {
       const meta = {
-        userID: rectifyUserID( directLine, userIDFromAction ),
+        userID: rectifyUserID(directLine, userIDFromAction),
         username
       };
 
       let endDirectLine;
 
-      yield put( { type: CONNECT_PENDING, meta } );
+      yield put({ type: CONNECT_PENDING, meta });
 
       try {
-        endDirectLine = yield call( connectSaga, directLine );
-      } catch ( err ) {
-        yield put( { type: CONNECT_REJECTED, error: true, meta, payload: err } );
+        endDirectLine = yield call(connectSaga, directLine);
+      } catch (err) {
+        yield put({ type: CONNECT_REJECTED, error: true, meta, payload: err });
 
         continue;
       }
 
-      // At this point, we established connection to Direct Line.
+      // At this point, we are establishing connection to Direct Line.
       // If there are any errors from this point, we need to make sure we call endDirectLine() to release resources.
       try {
-        yield put( { type: CONNECT_FULFILLING, meta, payload: { directLine } } );
-        yield put( { type: CONNECT_FULFILLED, meta, payload: { directLine } } );
-
-        const terminateAction = yield take( [ DISCONNECT, negativeUpdateConnectionStatusAction ] );
+        let gotDisconnectAction;
+
+        for (let numConnect = 0;; numConnect++) {
+          // This code section uses all() because we want to ensure CONNECT_FULFILLED is dispatched
+          // before we dispatch any connection-related actions, including DISCONNECT_FULFILLING and RECONNECT_PENDING.
+          const {
+            disconnection: {
+              takeDisconnectAction,
+              takeUpdateConnectionStatusAction
+            }
+          } = yield all(
+            {
+              _: numConnect ?
```

Comment:
it seems expected, but `numConnect` being 0 the first iteration is the only time this evaluates to false. I'm wondering if it get be written in a way to make it more clear. Maybe a comment denoting such and why would be helpful.

Reply:
Updated the code (in a big way). Could you please review it again?